### ISSUE TITLE
systemd v240 in debian buster Type=notify SuccessExitStatus still results in failure

### ISSUE THREAD

**Issue by daveolson53** [systemd/systemd#11627]
**systemd version the issue has been seen with**
> 240

**Used distribution**
 > Debian buster

**Expected behaviour you didn't see**
 > Exit value 5 should have been status Success

**Unexpected behaviour you saw**
 > Status failed

**Steps to reproduce the problem**
Exit with status in SuccessExitStatus prior to sending Notify message

**Comment by daveolson53** [comment]
This might be related to https://github.com/systemd/systemd/pull/4745.  If it is, I'm guessing it is an unintended side effect

**Comment by poettering** [comment]
so, we consider it a protocol violation if a service is configured for Type=notify and then exits before sending READY=1. because in that case the service is not properly started up after all.

What are you trying to do? If you are not sending READY=1 why are you using Type=notify? Not sure I grok the setup?

**Comment by daveolson53** [comment]
Indeed, the service has *not* started properly, and that's why we don't want it restarted on exit, and that's why we use SuccessExitStatus=5.

The situation is that the service is always enabled, but in some circumstances it is not configured, and therefore exits before the main event loop is entered.

As is normally the case with main event loops and noityf messages that send READY, the READY message isn't sent until initialization is complete and the main event loop is entered.

I'm not sure what the poit of SuccessExitStatus status is, if in some circumstances it is ignored, and that isn't documented anywhere.

And again, this worked up through at least systemd v232 so it's a break in compatibility that isn't documented in the commits, the man pages, or anywhere else.

**Comment by poettering** [comment]
I figure this behaviour was changed in c35755fb878af58b80ac62a501a75f79c90a3763 and related.

/cc @joukewitteveen 

The old behaviour was not documented either btw. We can certainly revisit this, but the fact that this worked and then stopped worked is in both case unsupported behaviour.

> I'm not sure what the point of SuccessExitStatus status is, if in some circumstances it is ignored, and that isn't documented anywhere.

Well, the thinking so far is: there are many many reasons why a service can fail. just some examples:

1. it aborts at any time
2. it fails to send READY=1 in time (timeout)
3. it violates its configured startup protocol, by exiting early without ever reporting a successfuly startup
4. it exits non-zero 4 days after it successfully completed start-up.
5. some kernel operation (for example fork() or setresuid()) fail for the service
…

Right now the SuccessExitStatus= setting is used to tweak case 4 of the above, but that's all. i.e. with the setting you can tweak that specific exit statuses seen after the service fully started up are handled as success. It however has no influence on case 3, i.e. where the service doesn't even manage to start-up but exits early...

I am not sure what the right approach here is in the long run. We definitely should start documenting beaviour here, that's without question, but the behaviour you'd like to see is a bit strange I think. I mean, a service with Type=notify that just does ExecStart=/bin/true and is accepted as "all is good" sounds strange to me, because quite obviously in this case the protocol chosen with type=notify and the protocol implemented in ExecStart= don't match at all, and it appears to me that should result in a failure state and not be silently accepted.

Anyway, @keszybz @yuwata @joukewitteveen ideas, opinions?

**Comment by joukewitteveen** [comment]
Could it be that what you want is to send `STOPPING=1` (you might need to send `READY=1` first)?
In my mind, a `Type=notify` service should communicate through notifications, not through exit status.

If this issue would indicate something that is currently not possible, I guess that means the notification `language' is not rich enough (yet).

**Comment by joukewitteveen** [comment]
Having read and thought a bit more, I think the following procedure should be, and likely already _is_, correct for what you try to achieve:
1. Send `STOPPING=1` (no need to send `READY=1` first)
2. Terminate with the intended exit status

**Comment by daveolson53** [comment]
The workaround I had come up with was to send READY and then exit; I can send STOPPING first, but it seems that might trigger some future problem if STOPPING is sent prior to READY.

I think the very restricted use of SuccessExitStatus is unnecessary, but I can accept it.

**Comment by joukewitteveen** [comment]
Doesn't sending `READY=1` and then exit leave the service active? Really, I believe what you want is to simply send `STOPPING=1` before you `exit 5` (and not, of course followed by `READY=1` later on).

**Comment by daveolson53** [comment]
Sending READY=1 and exiting leaves it as inactive(dead)  and Succeeded.

Sending STOPPING=1 and exitting causes a protocol failure and failed status.

**Comment by joukewitteveen** [comment]
I'd say the latter may be a bug.
